Align update of nextSIM-DG and XIOS calendars

joewallwork opened the pull request:
# Align update of nextSIM-DG and XIOS calendars

Fixes #756
Fixes #740
Refinement of #751

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

This PR adds functionality for incrementing XIOS' calendar and hooks it up in the `ModelMetadata::incrementTime` method.

More specifically:
* Define `Xios::incrementCalendar`.
* Rename `Xios::updateCalendar` and `Xios::setCalendarStep` (for consistency with `Xios::getCalendarStep`.
* Reconfigure the build system so that XIOS can be used more widely (i.e., in the library as well as the tests).

---
# Test Description

The new method is used in the XIOS read/write test. Unfortunately, I found a bug in XIOS (see comments below). However, this is fixed with a minor hack in the `install-xios.sh` script. We should be able to avoid this hack when we update to XIOS3 (see #761).

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

joewallwork committed:
Modify build system for wider-scale use of XIOS

joewallwork committed:
Add Xios::incrementCalendar and use in ModelMetadata::incrementTime

joewallwork committed:
Rename updateCalendar as setCalendarStep

joewallwork committed:
Use ModelMetadata::incrementTime in XIOS read/write test

joewallwork committed:
Create XIOS handlers for other tests build with XIOS support

joewallwork committed:
Same for ModelMetadata::setTime

joewallwork committed:
Enable XIOS; remember to close context defs

joewallwork committed:
Initialise xiosHandler to nullptr

joewallwork committed:
Don't build RectGrid_test with XIOS

joewallwork committed:
Add extra workflow job for testing without XIOS

joewallwork committed:
Pass timestep to Xios constructor when contextId is specified

joewallwork committed:
At the core of the issue in the read/write test

joewallwork committed:
Don't clear streams in Xios

joewallwork committed:
Fix ParaGrid_test

joewallwork committed:
Reset XIOS calendar step in ModelMetadata::setTime

joewallwork commented on core/test/XiosReadWrite_test.cpp:
```diff
@@ -128,6 +130,11 @@ Xios setupXiosHandler(int dim, bool read)
+    //        These are already set in the Xios constructor and the test fails if the following two
+    //        lines are moved above the close_context_definition.
+    xios_handler.setCalendarTimestep(Duration("P0-0T01:30:00"));
+    xios_handler.setCalendarStart(TimePoint("2023-03-17T17:11:00Z"));
```
I'm currently in discussion with the XIOS developers on why this is occurring. Apparently the MFE I provided works fine in XIOS3. (I'm using XIOS2 at revision 2638.)

joewallwork replied:
I considered updating to XIOS3 but found that some of the demos hang with that so figured it'd be better to address this specific issue first.

Fixed the issue above locally with a hack as in https://github.com/nextsimhub/nextsimdg/pull/757/commits/fe1c9b22c4e30054fd44040c2fccdd6ef8d181a2. My latest commit triggered the Docker image to be rebuilt so hopefully it will work in there, too.

joewallwork replied:
Okay so initially the test suite with XIOS failed but then once the Docker image had built, I re-triggered it and it passed!

joewallwork committed:
Issue fixed by XIOS hack

joewallwork committed:
Apply a hack to get XIOS2 working

joewallwork commented on .github/workflows/test_suite.yml:
```diff
@@ -50,9 +50,43 @@ jobs:
-        cd -
 
-  test-ubuntu-mpi:
+  test-ubuntu-mpi-noxios:
+
+    runs-on: ubuntu-22.04
+    container:
+      image: ghcr.io/nextsimhub/nextsimdg-dev-env:latest
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: build and compile with MPI but not XIOS
+      run: |
+        . /opt/spack-environment/activate.sh
+        mkdir -p build && cd build
+        cmake -DENABLE_MPI=ON -DENABLE_XIOS=OFF -DCMAKE_CXX_COMPILER="$(which mpic++)" ..
+        make -j 4
+
+    - name: run MPI tests
+      run: |
+        . /opt/spack-environment/activate.sh
+        apt update
+        apt install -y wget
+        cd build
+        (cd core/test && wget "ftp://ftp.nersc.no/nextsim/netCDF/partition_metadata_1.nc")
+        for component in core physics dynamics
+        do
+            cd $component/test
+            for file in $(find test* -maxdepth 0 -type f)
+            do
+              echo $file
+              nprocs=$(echo $file | sed -r "s/.*MPI([0-9]+)/\1/")
+              mpirun --allow-run-as-root --oversubscribe -n $nprocs ./$file
+            done
+            cd -
+        done
+
+  test-ubuntu-mpi-xios:
```
Note that this change is the same as the one in #758.

joewallwork commented on core/src/ParaGridIO.cpp:
```diff
@@ -227,7 +227,7 @@ ModelState ParaGridIO::readForcingTimeStatic(
         timeVar.getVar(timeVec.data());
         // Get the index of the largest TimePoint less than the target.
-        targetTIndex = std::find_if(begin(timeVec), end(timeVec), [time](double t) {
+        targetTIndex = std::find_if(std::begin(timeVec), std::end(timeVec), [time](double t) {
```
This is required to avoid compile errors about `begin` and `end` being overloaded and it not being clear which version to use.

joewallwork commented on core/src/Xios.cpp:
```diff
@@ -37,7 +37,6 @@ static const std::map<int, std::string> keyMap = { { Xios::ENABLED_KEY, "xios.en
 //! Enable XIOS in the 'config'
 void enableXios()
 {
-    Configurator::clearStreams();
```
This change is important when you set up XIOS after you've already configured other parts of nextSIM-DG.

joewallwork committed:
Drop FIXME no longer relevant

joewallwork committed:
assertIsClose no longer needed [skip ci]

joewallwork commented:
Oh nice - turns out the XIOS update fixes issue #740. (Dropped workaround in 28b24531.)

timspainNERSC commented on core/src/include/ModelMetadata.hpp:
```diff
@@ -117,6 +129,9 @@ class ModelMetadata {
     const std::string bboxName = "bounding_boxes";
 #endif
+#ifdef USE_XIOS
+    Xios* xiosHandler = nullptr;
```
Is `ModelMetadata` the best place for the XIOS handler? It doesn't (instinctively) feel like metadata data to me. Perhaps implementing the infrastructure necessary to return an `Xios` singleton.

Also, wherever it ends up, might a smarter pointer be better than a raw pointer?

joewallwork replied:
The way I currently have it set up in #751 is that multiple classes (`ParaGridIO` and `ModelMetadata`) have pointers to it. A singleton would make sense in general, yes. It might make test files containing multiple tests that use XIOS a little trickier to set up, but that's not really the important thing. I'll look into how to implement it. Also good point on using smart pointers.

joewallwork replied:
Opened PR #763 to address this issue, which would merge into this PR if approved.

TomMelt replied:
@timspainNERSC what kind of smart pointer were you considering? I guess we could do something like `unique_ptr` but this slightly different to the singleton functionality we need. We really want a single instance rather than a unique pointer. But perhaps both is extra protection?

timspainNERSC replied:
I think a `shared_ptr` is more likely to be the type of smart pointer that we need for a singleton.

timspainNERSC replied:
Thinking further, the singleton function should return a reference, rather than any kind of smart or dumb pointer. If an object needs to store a pointer to the object, then a plain pointer should be fine, as the ownership of the object is clear: it belongs to the singleton function.

TomMelt replied:
@AdelekeBankole , place the `unique_ptr` into `xios.hpp`

joewallwork replied:
This discussion will be resolved upon merging #763 with recent changes.

timspainNERSC reviewed (changes requested):
The XIOS handler living in `ModelMetadata` feels out of place. Perhaps define it as a singleton in `Xios.hpp`/`.cpp`?

joewallwork committed:
Define Xios handler as singleton

joewallwork committed:
Use singleton in most XIOS tests

joewallwork committed:
Use Xios singleton in other tests

joewallwork committed:
Recombine dims; re-separate read and write

joewallwork committed:
Enable XIOS in tests

joewallwork committed:
Revert date; drop unused mutex include

TomMelt reviewed (approved):
Once #763 is merged in this looks good to go :rocket:

joewallwork commented:
~~Let's merge this following #807.~~

Oops, wrong PR.

joewallwork committed:
Implement getInstance method as a reference to static Xios instance

joewallwork committed:
Remove added line [skip ci]

joewallwork committed:
Avoid setting defaults for constructor twice

joewallwork committed:
Prevent copying

joewallwork committed:
Fix reference usage [skip ci]

joewallwork committed:
Attempt at setting up finalizer

joewallwork committed:
Call finalizer in tests

joewallwork commented:
[Rebased on top of `develop`.]

joewallwork committed:
Set up `Finalizer` for XIOS (#822)

Supersedes #808.

As discussed in the meeting this morning, this PR hooks the `Xios` class
into the `Finalizer`. Since no `Model`s are created in the standalone
XIOS unit tests, they need to call `Finalizer::finalize` explicitly. It
seems the same is also true of the `ConfigOutput` and `ParaGrid` tests.
So actually the changeset isn't that different than @TomMelt's one in
#808.

joewallwork committed:
Implement Xios handler as a singleton (#763)

Note that this PR merges into the branch for #757, not `develop`.

This PR implements the `Xios` handler object as a singleton, as
requested. This means that nextSIM-DG classes don't need to own pointers
to the current working `Xios` instance. There can be only one `Xios`
instance. The first time the static `Xios::getInstance` method is
called, it creates the singleton and returns a pointer to it. On
subsequent calls, this method simply returns a pointer to the singleton.

Unfortunately, I had to rework the XIOS read/write test. The singleton
design pattern isn't suitable for the way it's currently written, so I
re-separated out the reading and writing parts and re-combined the 2D
and 3D cases.

joewallwork commented:
The same is true here as in #758 because of the (duplicate) workflow change.
> Ah, with this PR we have an issue that the repo settings expect the test-ubuntu-mpi action to pass, but it's been split out into test-ubuntu-mpi-noxios and test-ubuntu-mpi-xios. So we'll need to edit the repo settings in order to merge it.

timspainNERSC reviewed (approved):
That looks good, now

joewallwork commented:
> That looks good, now

Perfect, thanks @timspainNERSC. Please could you temporarily disable the requirement for the `test-ubuntu-mpi` test to pass before merging? Once this has been merged, there will then be `test-ubuntu-mpi-noxios` and `test-ubuntu-mpi-xios` jobs that we can require instead. (I guess I could alternatively rename `test-ubuntu-mpi-noxios` as `test-ubuntu-mpi` but I feel it's good to be explicit here.)

joewallwork committed:
Merge branch 'develop' into issue756_xios-calendar-increment

joewallwork commented:
Will merge once the periodic Docker build passes.
